jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/gridcolumnvisibility/JmixGridColumnVisibility.java

```diff
@@ -0,0 +1,653 @@
+import com.vaadin.flow.component.shared.HasOverlayClassName;
+import com.vaadin.flow.component.shared.HasThemeVariant;
+import com.vaadin.flow.shared.Registration;
+import io.jmix.core.Messages;
+import io.jmix.core.common.util.Preconditions;
+import io.jmix.flowui.component.grid.DataGrid;
+import io.jmix.flowui.component.grid.DataGridColumn;
+import io.jmix.flowui.component.grid.TreeDataGrid;
+import io.jmix.flowui.kit.component.HasTitle;
+import io.jmix.flowui.kit.component.dropdownbutton.AbstractDropdownButton;
+import io.jmix.flowui.kit.component.menubar.JmixMenuBar;
+import io.jmix.flowui.kit.component.menubar.JmixMenuItem;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.IntStream;
+
+/**
+ * A component which allows to manage visibility of data grid columns
+ */
+public class JmixGridColumnVisibility extends Composite<JmixMenuBar>
+        implements AttachNotifier, DetachNotifier, ApplicationContextAware, InitializingBean,
+        HasTitle, HasSize, HasThemeVariant<GridColumnVisibilityVariant>, HasEnabled, HasStyle, HasOverlayClassName,
+        HasText, Focusable<AbstractDropdownButton> {
+
+    protected static final String ATTRIBUTE_JMIX_ROLE_NAME = "jmix-role";
+    protected static final String ATTRIBUTE_JMIX_ROLE_VALUE = "jmix-grid-column-visibility";
+
+    protected ApplicationContext applicationContext;
+    protected Messages messages;
+
+    protected JmixMenuItem dropdownItem;
+    protected Icon icon;
+
+    protected Grid<?> grid;
+
+    protected Header header;
+    protected List<ColumnItemImpl> columnItems = new ArrayList<>();
+
+    public JmixGridColumnVisibility() {
+        dropdownItem = getContent().addItem("");
+    }
+
+    @Override
+    public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+        this.applicationContext = applicationContext;
+    }
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        autowireDependencies();
+        initHeader();
+    }
+
+    protected void autowireDependencies() {
+        messages = applicationContext.getBean(Messages.class);
+    }
+
+    protected void initHeader() {
+        header = new Header();
+        header.setShowAllEnabled(true);
+        header.setHideAllEnabled(true);
+    }
+
+    @Override
+    protected JmixMenuBar initContent() {
+        JmixMenuBar menuBar = super.initContent();
+        menuBar.getElement().setAttribute(ATTRIBUTE_JMIX_ROLE_NAME, ATTRIBUTE_JMIX_ROLE_VALUE);
+        return menuBar;
+    }
+
+    /**
+     * Sets component icon.
+     *
+     * @param icon icon to set
+     */
+    public void setIcon(@Nullable Icon icon) {
```

Comment:
Let's add `themeName`: `icon` automatically if only icon is present. Same as `com.vaadin.flow.component.button.Button#setIcon`.